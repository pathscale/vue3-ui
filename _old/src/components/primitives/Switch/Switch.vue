<script>
import { ref, watchEffect, computed } from 'vue'

export default {
  name: 'VSwitch',
  props: {
    modelValue: {
      type: [String, Number, Boolean, Function, Object, Array, Date],
    },
    nativeValue: {
      type: [String, Number, Boolean, Function, Object, Array, Date],
    },
    disabled: {
      type: Boolean,
      default: null
    },
    type: String,
    passiveType: String,
    name: String,
    required: Boolean,
    size: String,
    trueValue: {
      type: [String, Number, Boolean, Function, Object, Array, Date],
      default: true
    },
    falseValue: {
      type: [String, Number, Boolean, Function, Object, Array, Date],
      default: false
    },
    rounded: {
      type: Boolean,
      default: true
    },
    outlined: Boolean
  },
  emits: ['update:modelValue'],
  setup(props, { emit }) {
    const value = ref(props.modelValue)
    const passiveClass = computed(() => props.passiveType && `${props.passiveType}-passive`)
    watchEffect(() => {
      value.value = props.modelValue
    })
    watchEffect(() => {
      emit('update:modelValue', value.value)
    })
    return { value, passiveClass }
  }
}
</script>

<template>
  <label
    class="switch"
    :class="[
      size,
      {
        'is-disabled': disabled,
        'is-rounded': rounded,
        'is-outlined': outlined,
      }
    ]"
    :disabled="disabled">
    <input
      v-model="value"
      type="checkbox"
      @click.stop
      :disabled="disabled"
      :name="name"
      :required="required"
      :value="nativeValue"
      :true-value="trueValue"
      :false-value="falseValue" />
    <span
      class="check"
      :class="[passiveClass, type]" />
    <span class="control-label"><slot /></span>
  </label>
</template>
