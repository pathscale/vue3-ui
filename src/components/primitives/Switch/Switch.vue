<script setup lang="ts">
import { computed, ref, watchEffect } from "vue";

const props = withDefaults(
  defineProps<{
    // biome-ignore lint/suspicious/noExplicitAny: allow any type according to docs
    modelValue?: any;
    // biome-ignore lint/suspicious/noExplicitAny: allow any type according to docs
    nativeValue?: any;
    disabled?: boolean;
    type?:
      | "is-white"
      | "is-light"
      | "is-dark"
      | "is-black"
      | "is-text"
      | "is-primary"
      | "is-link"
      | "is-info"
      | "is-success"
      | "is-warning"
      | "is-danger";
    passiveType?:
      | "is-white"
      | "is-black"
      | "is-light"
      | "is-dark"
      | "is-primary"
      | "is-info"
      | "is-success"
      | "is-warning"
      | "is-danger"
      | string; // and any other colors you've set in the $colors list on Sass
    name?: string;
    required?: boolean;
    size?: "is-small" | "is-medium" | "is-large";
    // biome-ignore lint/suspicious/noExplicitAny: allow any type according to docs
    trueValue?: any;
    // biome-ignore lint/suspicious/noExplicitAny: allow any type according to docs
    falseValue?: any;
    rounded?: boolean;
    outlined?: boolean;
  }>(),
  {
    trueValue: true,
    falseValue: false,
    rounded: true,
  },
);

const emit = defineEmits(["update:modelValue"]);

const value = ref(props.modelValue);
const passiveClass = computed(
  () => props.passiveType && `${props.passiveType}-passive`,
);
watchEffect(() => {
  value.value = props.modelValue;
});
watchEffect(() => {
  emit("update:modelValue", value.value);
});
</script>

<template>
  <label
    class="switch"
    :class="[
      size,
      {
        'is-disabled': disabled,
        'is-rounded': rounded,
        'is-outlined': outlined,
      }
    ]"
    :disabled="disabled">
    <input
      v-model="value"
      type="checkbox"
      @click.stop
      :disabled="disabled"
      :name="name"
      :required="required"
      :value="nativeValue"
      :true-value="trueValue"
      :false-value="falseValue" />
    <span
      class="check"
      :class="[passiveClass, type]" />
    <span class="control-label"><slot /></span>
  </label>
</template>
