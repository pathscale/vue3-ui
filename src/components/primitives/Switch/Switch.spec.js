import { shallowMount } from '@vue/test-utils'
import Switch from './Switch.vue'

let wrapper

describe('Switch', () => {
  beforeEach(() => {
    wrapper = shallowMount(Switch)
  })

  it('is called', () => {
    expect(wrapper.exists()).toBeTruthy()
  })

  it('render correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })
})
