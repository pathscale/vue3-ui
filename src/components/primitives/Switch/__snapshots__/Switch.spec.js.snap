// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch render correctly 1`] = `<label class="switch is-rounded"><input type="checkbox" value="false" true-value="true" false-value="false"><span class="check"></span><span class="control-label"></span></label>`;
